### PR TITLE
feat: Rate-and-state friction with explicit time integration

### PR DESCRIPTION
**Pull request by VidarStiernstrom**
Adds rate-and-state friction kernels based on explicit time integration of the rate-and-state equations. Adds a quasi-dynamic solver with an embedded (adaptive) Runge-Kutta 3(2) method for rate/state.

### CODE REVIEW

File: src/coreComponents/physicsSolvers/contact/ContactFields.hpp [hunk 1]
```diff
@@ -74,6 +74,14 @@ DECLARE_FIELD( dispJump,
                WRITE_AND_READ,
                "Displacement jump vector in the local reference system" );
 
+DECLARE_FIELD( dispJump_n,
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/contact/ContactFields.hpp:77]: we should get rid of `old` at some point though.

File: src/coreComponents/physicsSolvers/inducedSeismicity/kernels/RateAndStateKernels.hpp [hunk 2]
```diff
@@ -109,21 +127,17 @@ class RateAndStateKernel
...
-    m_slipVelocity[k][0] = projectionScaling * m_traction[k][1];
-    m_slipVelocity[k][1] = projectionScaling * m_traction[k][2];
+    real64 const frictionCoefficient = m_frictionLaw.frictionCoefficient( k, m_slipRate[k], m_stateVariable[k] );
+    projectSlipRateBase(k, frictionCoefficient, m_shearImpedance, m_traction, m_slipRate, m_slipVelocity);
   }
 
   GEOS_HOST_DEVICE
-  camp::tuple< int, real64 > checkConvergence( StackVariables const & stack,
-                                               real64 const tol ) const
+  std::pair< int, real64 > checkConvergence( StackVariables const & stack,
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/inducedSeismicity/kernels/RateAndStateKernels.hpp]:
```suggestion
  camp::tuple< int, real64 > checkConvergence( StackVariables const & stack,
```

File: src/coreComponents/physicsSolvers/inducedSeismicity/kernels/RateAndStateKernels.hpp [hunk 3]
```diff
@@ -109,21 +127,17 @@ class RateAndStateKernel
...
+                                             real64 const tol ) const
   {
     real64 const residualNorm = LvArray::tensorOps::l2Norm< 2 >( stack.rhs );
     int const converged = residualNorm < tol ? 1 : 0;
-    camp::tuple< int, real64 > result { converged, residualNorm };
-    return result;
+    return std::make_pair( converged, residualNorm );
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/inducedSeismicity/kernels/RateAndStateKernels.hpp]:
```suggestion
    camp::tuple< int, real64 > result { converged, residualNorm };
    return result;
```

File: src/coreComponents/physicsSolvers/inducedSeismicity/QuasiDynamicEQRK32.hpp [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+   */
+  virtual string getCatalogName() const override { return catalogName(); }
+
+  /// This method ties properties with their supporting mesh
+  virtual void registerDataOnMesh( Group & meshBodies ) override;
+
+  struct viewKeyStruct : public PhysicsSolverBase::viewKeyStruct
+  {
+    /// stress solver name
+    static constexpr char const * stressSolverNameString() { return "stressSolverName"; }
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/inducedSeismicity/QuasiDynamicEQRK32.hpp]:
```suggestion
     constexpr static char const * stressSolverNameString() { return "stressSolverName"; }
```

File: src/coreComponents/physicsSolvers/inducedSeismicity/kernels/RateAndStateKernels.hpp [hunk 5]
```diff
@@ -202,8 +321,258 @@ createAndLaunch( SurfaceElementSubRegion & subRegion,
...
+};
+
+/**
+ * @brief Runge-Kutta method used to time integrate slip and state. Uses of a high order
+ * update used to integrate the solutions, and a lower order update to estimate the error
+ * in the time step.
+ * 
+ * @tparam Butcher table defining the Runge-Kutta method. 
+ */
+template <typename Table> class EmbeddedRungeKuttaKernel
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/inducedSeismicity/kernels/RateAndStateKernels.hpp]:
```suggestion
template <typename TABLE_TYPE> class EmbeddedRungeKuttaKernel
```

File: src/coreComponents/physicsSolvers/inducedSeismicity/kernels/RateAndStateKernels.hpp [hunk 6]
```diff
@@ -202,8 +321,258 @@ createAndLaunch( SurfaceElementSubRegion & subRegion,
...
+{   
+    integer const algHighOrder = 3;                    // High-order update order 
+    integer const algLowOrder = 2;                     // Low-order update order
+    integer const numStages = 3;                       // Number of stages
+    real64 const a[2][2] = { { 1.0/2.0, 0.0 },         // Coefficients for stage value updates
+                             { -1.0,    2.0 } };       // (lower-triangular part of table).
+    real64 const c[3] = { 0.0, 1.0/2.0, 1.0 };         // Coefficients for time increments of substages
+    real64 const b[3] = { 1.0/6.0, 4.0/6.0, 1.0/6.0 }; // Quadrature weights used to step the solution to next time
+    real64 const bStar[3] = { 1.0/2.0, 0.0, 1.0/2.0 }; // Quadrature weights used for low-order comparision solution
+    real64 const FSAL = false;                         // Not first same as last
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/inducedSeismicity/kernels/RateAndStateKernels.hpp]: these can all be made `constexpr`.

**Review by CusiniM — APPROVED** [review]
It's a good start. You should run `make uncrustify_style` to fix the formatting. 

I have created an issue to keep track of things that need to be improved.

https://github.com/GEOS-DEV/GEOS/issues/3482

File: src/coreComponents/physicsSolvers/inducedSeismicity/QuasiDynamicEQRK32.hpp (hunk 4 above)
Comment (VidarStiernstrom) [reply src/coreComponents/physicsSolvers/inducedSeismicity/QuasiDynamicEQRK32.hpp]: Fixed

File: src/coreComponents/physicsSolvers/inducedSeismicity/kernels/RateAndStateKernels.hpp (hunk 6 above)
Comment (VidarStiernstrom) [reply src/coreComponents/physicsSolvers/inducedSeismicity/kernels/RateAndStateKernels.hpp]: I added this.

File: inputFiles/inducedSeismicity/inducedSeismicity.ats [hunk 7]
```diff
@@ -28,6 +28,13 @@ decks = [
         partitions=((1, 1, 1), ),
         restart_step=0,
         check_step=3262,
-        restartcheck_params=RestartcheckParameters(atol=1e-4, rtol=1e-3)) 
+        restartcheck_params=RestartcheckParameters(atol=1e-4, rtol=1e-3))
```
Comment (CusiniM) [inline inputFiles/inducedSeismicity/inducedSeismicity.ats]:
```suggestion
        restartcheck_params=RestartcheckParameters(atol=1e-4, rtol=1e-3)),
```

**Pull request merged**